Lib/multiprocessing/shared_memory.py

```diff
@@ -1,225 +1,250 @@
-"Provides shared memory for direct access across processes."
+"""Provides shared memory for direct access across processes.
+
+The API of this package is currently provisional. Refer to the
+documentation for details.
+"""
 
 
 __all__ = [ 'SharedMemory', 'PosixSharedMemory', 'WindowsNamedSharedMemory',
-            'ShareableList', 'shareable_wrap',
-            'SharedMemoryServer', 'SharedMemoryManager', 'SharedMemoryTracker' ]
+            'ShareableList',
+            'SharedMemoryServer', 'SharedMemoryManager' ]
 
 
-from functools import reduce
+from functools import partial
 import mmap
-from .managers import DictProxy, SyncManager, Server
+from .managers import dispatch, BaseManager, Server, State, ProcessError
 from . import util
 import os
-import random
 import struct
-import sys
+import secrets
 try:
-    from _posixshmem import _PosixSharedMemory, Error, ExistentialError, O_CREX
+    import  _posixshmem
+    from os import O_CREAT, O_EXCL, O_TRUNC
 except ImportError as ie:
-    if os.name != "nt":
-        # On Windows, posixshmem is not required to be available.
-        raise ie
-    else:
-        _PosixSharedMemory = object
-        class ExistentialError(BaseException): pass
-        class Error(BaseException): pass
-        O_CREX = -1
+    O_CREAT, O_EXCL, O_TRUNC = 64, 128, 512
+
+O_CREX = O_CREAT | O_EXCL
+
+if os.name == "nt":
+    import _winapi
+
+    PAGE_READONLY = 0x02
+    PAGE_EXECUTE_READWRITE = 0x04
+    INVALID_HANDLE_VALUE = -1
 
 
 class WindowsNamedSharedMemory:
 
-    def __init__(self, name, flags=None, mode=None, size=None, read_only=False):
+    def __init__(self, name, flags=None, mode=384, size=0, read_only=False):
         if name is None:
-            name = f'wnsm_{os.getpid()}_{random.randrange(100000)}'
+            name = _make_filename()
+
+        if size == 0:
+            # Attempt to dynamically determine the existing named shared
+            # memory block's size which is likely a multiple of mmap.PAGESIZE.
+            try:
+                h_map = _winapi.OpenFileMappingW(PAGE_READONLY, False, name)
```

Comment:
This should request `FILE_MAP_READ` (4) access. `PAGE_READONLY` (2) is a page protection for `CreateFileMapping`. It happens to be the same constant value as `FILE_MAP_WRITE` access. 

`CreateFileMapping` determines the desired access for the section object from the given page protection. OTOH, with `OpenFileMapping` we have to supply the desired access explicitly.

Reply:
Thanks for catching this.  Changed to ``_winapi.FILE_MAP_READ`` in commit 68785333a25d902a42252ace3a8268357efd6094.